src/errorpage.cc

```diff
@@ -810,67 +812,59 @@ ErrorState::~ErrorState()
+         "ErrPage: " << errorPageName(type) << "\r\n";
+
+    if (xerrno)
+        body << "Err: (" << xerrno << ") " << strerror(xerrno) << "\r\n";
+    else
+        body << "Err: [none]\r\n";
 
-    str.reset();
-    /* email subject line */
-    str.appendf("CacheErrorInfo - %s", errorPageName(type));
-    mb->appendf("?subject=%s", rfc1738_escape_part(str.buf));
-    str.reset();
-    /* email body */
-    str.appendf("CacheHost: %s\r\n", getMyHostname());
-    /* - Err Msgs */
-    str.appendf("ErrPage: %s\r\n", errorPageName(type));
-
-    if (xerrno) {
-        str.appendf("Err: (%d) %s\r\n", xerrno, strerror(xerrno));
-    } else {
-        str.append("Err: [none]\r\n", 13);
-    }
 #if USE_AUTH
     if (auth_user_request.getRaw() && auth_user_request->denyMessage())
-        str.appendf("Auth ErrMsg: %s\r\n", auth_user_request->denyMessage());
+        body << "Auth ErrMsg: " << auth_user_request->denyMessage() << "\r\n";
 #endif
+
     if (dnsError)
-        str.appendf("DNS ErrMsg: " SQUIDSBUFPH "\r\n", SQUIDSBUFPRINT(*dnsError));
+        body << "DNS ErrMsg: " << *dnsError << "\r\n";
 
-    /* - TimeStamp */
-    str.appendf("TimeStamp: %s\r\n\r\n", Time::FormatRfc1123(squid_curtime));
+    body << "TimeStamp: " << Time::FormatRfc1123(squid_curtime) <<
+        "\r\n\r\n";
 
-    /* - IP stuff */
-    str.appendf("ClientIP: %s\r\n", src_addr.toStr(ntoabuf,MAX_IPSTRLEN));
+    char ntoabuf[MAX_IPSTRLEN];
+    body << "ClientIP: " << src_addr.toStr(ntoabuf, MAX_IPSTRLEN) << "\r\n";
 
-    if (request && request->hier.host[0] != '\0') {
-        str.appendf("ServerIP: %s\r\n", request->hier.host);
-    }
+    if (request && request->hier.host[0] != '\0')
+        body << "ServerIP: " << request->hier.host << "\r\n";
+    body << "\r\n";
+
+    body << "HTTP Request:\r\n";
 
-    str.append("\r\n", 2);
-    /* - HTTP stuff */
-    str.append("HTTP Request:\r\n", 15);
     if (request) {
-        str.appendf(SQUIDSBUFPH " " SQUIDSBUFPH " %s/%d.%d\n",
-                    SQUIDSBUFPRINT(request->method.image()),
-                    SQUIDSBUFPRINT(request->url.path()),
-                    AnyP::ProtocolType_str[request->http_ver.protocol],
-                    request->http_ver.major, request->http_ver.minor);
-        request->header.packInto(&str);
+            MemBuf r;
+            r.init();
+            request->pack(&r);
+            body << r.content();
     }
 
-    str.append("\r\n", 2);
     /* - FTP stuff */
 
     if (ftp.request) {
-        str.appendf("FTP Request: %s\r\n", ftp.request);
-        str.appendf("FTP Reply: %s\r\n", (ftp.reply? ftp.reply:"[none]"));
-        str.append("FTP Msg: ", 9);
-        wordlistCat(ftp.server_msg, &str);
-        str.append("\r\n", 2);
+        body << "FTP Request: " << ftp.request << "\r\n" <<
+            "FTP Reply: " << (ftp.reply ? ftp.reply : "[none]") << "\r\n" <<
+            "FTP Msg: " << AsList(*(ftp.server_msg)) << "\r\n\r\n";
```

Comment:
For safety sake and consistency with `%g` code:

```suggestion
            "FTP Reply: " << (ftp.reply ? ftp.reply : "[none]") << "\r\n";
        if (ftp.server_msg)
            body << "FTP Msg: " << AsList(*ftp.server_msg) << "\r\n";
        body << "\r\n";
```

To ease reading, I also removed excessive parenthesis in the above suggestion.

N.B. The removed wordlistCat()-based code offered similar nil pointer protections (but ended up printing an arguably misleading `FTP Msg:` prefix).

Reply:
done